server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1596,6 +1608,37 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
         }
     }
 
+    /**
+     * This method tries to rewrite a query without using a searcher (before we get a valid one). It takes advantage of the fact that
+     * we can skip some shards in the query phase because we have enough information in the index mapping to decide the 'can match'
+     * outcome. One such example are queries like {@link org.elasticsearch.index.query.MatchPhraseQueryBuilder} queries which are rewritten
+     * to {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} after being rewritten to
+     * {@link org.elasticsearch.index.query.TermQueryBuilder}. Other queries have a similar pattern, especially when using
+     * {@link org.elasticsearch.index.mapper.ConstantFieldType} or unmapped fields, they all rewrite to
+     * {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} (if conditions are met). This allows us to avoid extra work other than
+     * making the shard search active and waiting for refreshes. As a result, we only wait for refreshes to happen on shards that have
+     * actual data.
+     */
+    private static CanMatchShardResponse canMatchAfterRewriteWithoutSearcher(
+        final ShardSearchRequest request,
+        final IndexService indexService
+    ) throws IOException {
+        final SearchExecutionContext searchExecutionContext = new SearchExecutionContext(
+            indexService.newSearchExecutionContext(
+                request.shardId().getId(),
+                request.shardRequestIndex(),
+                null,
```

Comment:
I wonder if passing a null searcher warrants a cleanup in a followup. I know it's a thing that may already happen but I am not sure we should build upon that further. The rewrite needs a `QueryRewriteContext`, which though will call `convertToSearchExecutionContext` to call `getFieldType` against it. Maybe the query rewrite context could expose  `getFieldType` that would work when mappings are available (in the data node that has the index), to cover for situations where we want to rewrite purely based on mappings like in this case. I see though that `ConstantFieldType#matches` requires the `SearchExecutionContext` as an argument so the methods needed are more than just `getFieldType` and the change I am suggesting is more involving than it may seem... @romseygeek double checking if you have ideas

Reply:
I think it's just `getFieldType` and `getIndexSettings()`, both of which can be happily moved into `QueryRewriteContext` I think?  Ideally we could move a couple of more methods in there and remove the abomination that is `convertToSearchExecutionContext()`....

Reply:
And `indexNamedMatcher`.  And actually we can move `getWriteableRegistry` and `nowInMillis` into `SearchExecutionContext` as they are never used for rewriting, which nearly balances things out.

Reply:
@romseygeek so if I understand correctly the idea is to move all of that stuff into `QueryRewriteContext` so we don't need a `SearchExecutionContext` at all?

Reply:
Yes, precisely.  I'm working on a PR now...

Reply:
I'd advocate for the needed refactoring to be a prerequisite of the PR. I think it does not make a lot of sense to me to build on the search context with null searcher if we want to rather rely on a cleaner approach that relies on the query rewrite context. Otherwise we add in tech debt that we are going to have to remove straight-away or will stay around for much longer if we don't get to it? What do you folks think?